### PR TITLE
replaced insert into sent query with helper_sent.insert method and al…

### LINKED ISSUE #1524: uuid instead of empty msgid for new messages in sent fold
newly created messages have a empty msgid, which is later updated by the worker thread. This creates problems. I recommend using `uuid.uuid4().bytes` instead, which will again be replaced by the worker thread later. This can probably be done in `helper_sent` (except bitmessageqt which doesn't use that, so that needs to be refactord to use helper_sent). The signals the worker thread sends back should also be modified to send the old msgid if it differs from the new msgid, and not to use `updateSentItemStatusByToAddress` anymore.

The interfaces (curses, qt, api, kivy) need to be adjusted accordingly, so that they can replace the msgid if they get a new one from the signal.

- [ ] all interfaces that insert into sent table should use `helper_sent.insert`
  - [ ] Qt
  - [ ] kivy
  - [ ] curses
- [ ] `helper_sent.insert` should check the first argument is empty or None, and if yet, replace it with uuid, and return said uuid (otherwise return the first argument)
- [ ] the UI should retrieve this UUID and adjust the UI accordingly
  - [ ] Qt
  - [ ] kivy
  - [ ] curses

### PR DESCRIPTION
**Pull request by navjotcis**
1) I have replaced insert into sent query execution syntax with the helper_sent.insert method.
2) I have also replaced the empty msgid with UUID which will later be updated by a worker thread.
3) fixes #1524

### CODE REVIEW

File: src/bitmessagecurses/__init__.py [hunk 1]
```diff
@@ -31,6 +31,8 @@
 from helper_ackPayload import genAckPayload
 from helper_sql import sqlExecute, sqlQuery
 from inventory import Inventory
+
+import helper_sent
```
Comment (PeterSurda) [inline src/bitmessagecurses/__init__.py:35]: wrong order, should be between lines 21 and 22

File: src/bitmessagecurses/__init__.py [hunk 2]
```diff
@@ -982,7 +983,9 @@ def sendMessage(sender="", recv="", broadcast=None, subject="", body="", reply=F
                         0,      # retryNumber
                         "sent",
                         2,      # encodingType
-                        BMConfigParser().getint('bitmessagesettings', 'ttl'))
+                        BMConfigParser().getint('bitmessagesettings', 'ttl')
```
Comment (PeterSurda) [inline src/bitmessagecurses/__init__.py:986]: I would move this variable into `helper_sent` as well. Since you're already working on the code, may as well do it.

File: src/helper_sent.py [hunk 3]
```diff
@@ -2,9 +2,14 @@
 Insert values into sent table
 """
 
+import uuid
 from helper_sql import sqlExecute
 
 
 def insert(t):
     """Perform an insert into the `sent` table"""
+    msgid = uuid.uuid4().bytes
```
Comment (PeterSurda) [inline src/helper_sent.py:11]: only `if not t[0]:`

**Review by PeterSurda — CHANGES_REQUESTED** [review]
Please make the requested changes.

File: src/bitmessagecurses/__init__.py (hunk 1 above)
Comment (navjotcis) [reply src/bitmessagecurses/__init__.py:35]: ok

File: src/helper_sent.py (hunk 3 above)
Comment (navjotcis) [reply src/helper_sent.py:11]: ok

File: src/bitmessagecurses/__init__.py (hunk 2 above)
Comment (navjotcis) [reply src/bitmessagecurses/__init__.py]: ok

File: src/helper_sent.py [hunk 4]
```diff
@@ -2,9 +2,16 @@
...
+from bmconfigparser import BMConfigParser
 from helper_sql import sqlExecute
 
 
 def insert(t):
     """Perform an insert into the `sent` table"""
+    temp = list(t)
+    if not t[0]:
+        temp[0] = uuid.uuid4().bytes    # if msgid is empty the put uuid
+    temp[-1] = BMConfigParser().getint('bitmessagesettings', 'ttl')
```
Comment (PeterSurda) [inline src/helper_sent.py:15]: This should be optional.

File: src/helper_sent.py [hunk 5]
```diff
@@ -2,9 +2,16 @@
...
+    t = tuple(temp)
```
Comment (PeterSurda) [inline src/helper_sent.py:16]: Why is this necessary?

File: src/helper_sent.py [hunk 6]
```diff
@@ -2,9 +2,16 @@
...
+import uuid
+from bmconfigparser import BMConfigParser
 from helper_sql import sqlExecute
 
 
 def insert(t):
     """Perform an insert into the `sent` table"""
+    temp = list(t)
+    if not t[0]:
+        temp[0] = uuid.uuid4().bytes    # if msgid is empty the put uuid
```
Comment (PeterSurda) [inline src/helper_sent.py:14]: ... empty **then** put ...

Comment (PeterSurda) [inline src/helper_sent.py:14]: Needs to be tested regarding sqlite data type. Write a test for this.

**Review by PeterSurda — CHANGES_REQUESTED** [review]

File: src/helper_sent.py [hunk 7]
```diff
@@ -2,9 +2,20 @@
...
+from bmconfigparser import BMConfigParser
 from helper_sql import sqlExecute
 
 
 def insert(t):
     """Perform an insert into the `sent` table"""
+    if not t[0] or not t[-1]:
+        temp = list(t)
+        if not t[0]:
+            print('zero index is not available')
```
Comment (PeterSurda) [inline src/helper_sent.py:15]: no `print`

File: src/helper_sent.py [hunk 8]
```diff
@@ -2,9 +2,20 @@
...
+            temp[0] = uuid.uuid4().bytes    # if msgid is empty the put uuid
+        if not t[-1]:
+            print('Lasr index is not available')
```
Comment (PeterSurda) [inline src/helper_sent.py:18]: no `print`

**Review by PeterSurda — CHANGES_REQUESTED** [review]
Remove `print`.

Write a unit tests that verifies that insert is working correctly, in particular look out that `uuid.uuid4().bytes` needs to be recorded in the database as the correct data type after retrieving it.

**Review by PeterSurda — CHANGES_REQUESTED** [review]
no prints, add defaults, improve test

File: src/helper_sent.py [hunk 9]
```diff
@@ -10,38 +10,41 @@
 from helper_sql import sqlExecute
 
 
-def insert(t, is_testcase=False):
+# pylint: disable=too-many-arguments
+def insert(msgid, toAddress, ripe, fromAddress, subject, message, ackdata,
+           sentTime, lastActionTime, sleeptill=0, status='msgqueued',
+           retryNumber=0, folder='sent', encoding=2, ttl=0, is_testcase=False):
```
Comment (PeterSurda) [inline src/helper_sent.py:16]: default values missing

File: src/bitmessageqt/__init__.py [hunk 10]
```diff
@@ -2221,23 +2220,23 @@ def click_pushButtonSend(self):
...
-                     int(time.time()), # lastActionTime
-                     0, # sleepTill time. This will get set when the POW gets done.
-                     'broadcastqueued', 
-                     0, # retryNumber
-                     'sent', # folder
-                     encoding, # encoding type
-                     0
-                     )
-                helper_sent.insert(t)
+                helper_sent.insert(
```
Comment (PeterSurda) [inline src/bitmessageqt/__init__.py:2223]: should use named arguments, defaults should be omitted

**Review by PeterSurda — CHANGES_REQUESTED** [review]
The helper function arguments need to have defaults, and calling it should only specify those arguments that aren't default.

File: src/helper_sent.py (hunk 9 above)
Comment (navjotcis) [reply src/helper_sent.py:16]: which values are missing?

File: src/bitmessagecurses/__init__.py [hunk 11]
```diff
@@ -992,24 +979,9 @@ def sendMessage(sender="", recv="", broadcast=None, subject="", body="", reply=F
             # dummy ackdata, no need for stealth
             ackdata = genAckPayload(decodeAddress(addr)[2], 0)
```
Comment (PeterSurda) [inline src/bitmessagecurses/__init__.py:980]: can be removed now

File: src/helper_sent.py [hunk 12]
```diff
@@ -2,9 +2,44 @@
...
+
+    status = status if status else 'msgqueued'
+    ttl = ttl if ttl else BMConfigParser().getint('bitmessagesettings', 'ttl')
+
+    t = (msgid, toAddress, ripe, fromAddress, subject, message, ackdata,
+         sentTime, lastActionTime, sleeptill, status, retryNumber, folder,
+         encoding, ttl)
+
     sqlExecute('''INSERT INTO sent VALUES (?,?,?,?,?,?,?,?,?,?,?,?,?,?,?)''', *t)
+    return msgid
```
Comment (PeterSurda) [inline src/helper_sent.py:45]: I changed my mind here, it should return `ackdata` and not `msgid`, based on reviewing existing code.

File: src/tests/core.py [hunk 13]
```diff
@@ -238,6 +240,19 @@ def test_version(self):
...
+        fromAddress = 'BM-2cTrmD22fLRrumi3pPLg1ELJ6PdAaTRTdfg'
+        toAddress = 'BM-2cVWtdUzPwF7UNGDrZftWuHWgjdfkj89fdf'
+        message = 'test message'
+        subject = 'test subject'
+        status = 'msgqueued'
+        result = helper_sent.insert(
+            toAddress=toAddress, fromAddress=fromAddress, subject=subject,
+            message=message, status=status,
+        )
+        self.assertNotEqual(result, '')
```
Comment (PeterSurda) [inline src/tests/core.py:254]: It should retrieve the row from the sent table based on `result` and do some check on the values.

File: src/class_objectProcessor.py [hunk 14]
```diff
@@ -749,28 +749,14 @@ def processmsg(self, data):
 
                 ackdata = genAckPayload(streamNumber, 0)
                 toAddress = '[Broadcast subscribers]'
```
Comment (PeterSurda) [inline src/class_objectProcessor.py:751]: these two lines can be removed.

File: src/helper_sent.py [hunk 15]
```diff
@@ -2,9 +2,44 @@
...
+           status=None, ripe=None, ackdata=None, sentTime=None, lastActionTime=None,
+           sleeptill=0, retryNumber=0, encoding=2, ttl=None, folder='sent'):
     """Perform an insert into the `sent` table"""
+    # pylint: disable=unused-variable
+    # pylint: disable-msg=too-many-locals
+
+    msgid = msgid if msgid else uuid.uuid4().bytes
+
+    if not ripe or not ackdata:
+        new_status, addressVersionNumber, streamNumber, new_ripe = decodeAddress(toAddress)
```
[the remark this answers is not in the document]
Comment (PeterSurda) [reply src/helper_sent.py:24]: It should check if it's a broadcast either based on `status`, or on missing `toAddress`.

File: src/helper_sent.py [hunk 16]
```diff
@@ -2,9 +2,44 @@
...
+import uuid
+from addresses import decodeAddress
+from bmconfigparser import BMConfigParser
+from helper_ackPayload import genAckPayload
 from helper_sql import sqlExecute
 
 
-def insert(t):
+# pylint: disable=too-many-arguments
+def insert(msgid=None, toAddress=None, fromAddress=None, subject=None, message=None,
```
Comment (PeterSurda) [inline src/helper_sent.py:14]: default `toAddress` should be `'[Broadcast subscribers]'`

**Review by PeterSurda — CHANGES_REQUESTED** [review]
Remove statements from the calling methods that were moved to `insert` (`ripe`, `genAckPayload`, ...).

File: src/helper_sent.py (hunk 12 above)
Comment (navjotcis) [reply src/helper_sent.py:45]: ok

File: src/class_objectProcessor.py (hunk 14 above)
Comment (navjotcis) [reply src/class_objectProcessor.py:751]: we can not remove line 751 because it is used in below lines

Comment (PeterSurda) [reply src/class_objectProcessor.py:751]: ok

File: src/api.py [hunk 17]
```diff
@@ -1174,24 +1161,11 @@ def HandleSendBroadcast(
         streamNumber = decodeAddress(fromAddress)[2]
```
Comment (PeterSurda) [inline src/api.py:1161]: remove

File: src/tests/core.py [hunk 18]
```diff
@@ -238,6 +241,20 @@ def test_version(self):
...
+        toAddress = 'BM-2cVWtdUzPwF7UNGDrZftWuHWgjdfkj89fdf'
+        message = 'test message'
+        subject = 'test subject'
+        result = helper_sent.insert(
+            toAddress=toAddress, fromAddress=fromAddress,
+            subject=subject, message=message,
+        )
+        queryreturn = sqlQuery(
+            '''select msgid from sent where ackdata=?''', result)
+        self.assertNotEqual(queryreturn[0][0], '')
```
Comment (PeterSurda) [inline src/tests/core.py:256]: test the data type for ackdata as well.

**Review by PeterSurda — CHANGES_REQUESTED** [review]

**Review by PeterSurda — APPROVED** [review]
